### PR TITLE
add anyio.Sequencer based on trio.testing.Sequencer

### LINKED ISSUE #261: add anyio.testing.Sequencer to mirror trio.testing.Sequencer

### PR DESCRIPTION
**Pull request by graingert**
Fixes #261

### CODE REVIEW

File: src/anyio/_backends/_trio.py [hunk 1]
```diff
@@ -721,6 +726,24 @@ def wait_all_tasks_blocked():
     return trio.testing.wait_all_tasks_blocked()
 
 
+def _sequencer_factory() -> "trio_testing.Sequencer":
+    import trio.testing
+
```
Comment (graingert) [inline src/anyio/_backends/_trio.py:731]:
```suggestion
```

File: src/anyio/_core/_testing.py [hunk 2]
```diff
@@ -65,3 +65,43 @@ def get_running_tasks() -> DeprecatedAwaitableList[TaskInfo]:
...
+             async with seq(5):
+                 print(5)
+         async def worker3(seq):
+             async with seq(1):
+                 print(1)
+             async with seq(3):
+                 print(3)
+         async def main():
+            seq = anyio.testing.Sequencer()
+            async with aynio.create_task_group() as task_group:
```
Comment (graingert) [inline src/anyio/_core/_testing.py:97]:
```suggestion
            async with anyio.create_task_group() as task_group:
```

File: src/anyio/_core/_testing.py [hunk 3]
```diff
@@ -65,3 +65,43 @@ def get_running_tasks() -> DeprecatedAwaitableList[TaskInfo]:
...
+                 print(2)
+             async with seq(5):
+                 print(5)
+         async def worker3(seq):
+             async with seq(1):
+                 print(1)
+             async with seq(3):
+                 print(3)
+         async def main():
+            seq = anyio.testing.Sequencer()
```
Comment (graingert) [inline src/anyio/_core/_testing.py:96]:
```suggestion
            seq = anyio.Sequencer()
```

File: src/anyio/_backends/_asyncio.py [hunk 4]
```diff
@@ -1813,3 +1818,39 @@ def exception_handler(loop: asyncio.AbstractEventLoop, context: Dict[str, Any])
             raise ExceptionGroup(exceptions)
 
         return retval
+
+
+@dataclass(eq=False)
+class Sequencer(BaseSequencer):
+    # lifted from https://git.io/JOP0M
```
Comment (graingert) [inline src/anyio/_backends/_asyncio.py:1825]:
```suggestion
    # lifted from trio.testing.Sequencer https://git.io/JOP0M
```

**Comment by agronholm** [comment]
This still needs the refactorings we talked about (no backend specific classes).

File: src/anyio/_backends/_asyncio.py [hunk 5]
```diff
@@ -56,6 +59,8 @@ def find_root_task() -> asyncio.Task:
         raise RuntimeError('Cannot find root task for setting cleanup callback')
 else:
 
+    from async_generator import asynccontextmanager
```
Comment (graingert) [inline src/anyio/_backends/_asyncio.py:62]: contextlib2

**Pull request closed without merging**